pkg/oauthserver/authenticator/password/keystonepassword/keystonepassword.go

```diff
@@ -16,24 +18,116 @@ import (
 type keystonePasswordAuthenticator struct {
-	providerName   string
-	url            string
-	client         *http.Client
-	domainName     string
-	identityMapper authapi.UserIdentityMapper
+	providerName           string
+	url                    string
+	client                 *http.Client
+	domainName             string
+	identityMapper         authapi.UserIdentityMapper
+	useKeystoneIdentity    bool
 }
 
 // New creates a new password authenticator that uses OpenStack keystone to authenticate a user by password
 // A custom transport can be provided (typically to customize TLS options like trusted roots or present a client certificate).
 // If no transport is provided, http.DefaultTransport is used
-func New(providerName string, url string, transport http.RoundTripper, domainName string, identityMapper authapi.UserIdentityMapper) authenticator.Password {
+func New(providerName string, url string, transport http.RoundTripper, domainName string, identityMapper authapi.UserIdentityMapper, useKeystoneIdentity bool) authenticator.Password {
 	if transport == nil {
 		transport = http.DefaultTransport
 	}
 	client := &http.Client{Transport: transport}
-	return &keystonePasswordAuthenticator{providerName, url, client, domainName, identityMapper}
+	return &keystonePasswordAuthenticator{providerName, url, client, domainName, identityMapper, useKeystoneIdentity}
+}
+
+// Authenticate user and return his ID from Keystone
+func getUserIDv3(client *gophercloud.ProviderClient, options gophercloud.AuthOptions) (string, error) {
+	// Override the generated service endpoint with the one returned by the version endpoint.
+	v3Client := openstack.NewIdentityV3(client)
+
+	// copy the auth options to a local variable that we can change. `options`
+	// needs to stay as-is for reauth purposes
+	v3Options := options
+
+	var scope *tokens3.Scope
+	if options.TenantID != "" {
+		scope = &tokens3.Scope{
+			ProjectID: options.TenantID,
+		}
+		v3Options.TenantID = ""
+		v3Options.TenantName = ""
+	} else {
+		if options.TenantName != "" {
+			scope = &tokens3.Scope{
+				ProjectName: options.TenantName,
+				DomainID:    options.DomainID,
+				DomainName:  options.DomainName,
+			}
+			v3Options.TenantName = ""
+		}
+	}
+
+	result := tokens3.Create(v3Client, v3Options, scope)
+
+	token, err := result.ExtractToken()
+	if err != nil {
+		return "", err
+	}
+
+	catalog, err := result.ExtractServiceCatalog()
+	if err != nil {
+		return "", err
+	}
+
+	// Parse the body to obtain userID from the response
+	out, err := json.Marshal(result.Body)
+	if err != nil {
+		return "", err
+	}
+
+	var bd keystoneResponseBody
+
+	err = json.Unmarshal(out, &bd)
+	if err != nil {
+		return "", err
+	}
+
+	client.TokenID = token.ID
+
+	if options.AllowReauth {
+		client.ReauthFunc = func() error {
+			client.TokenID = ""
+			return openstack.AuthenticateV3(client, options)
+		}
+	}
+	client.EndpointLocator = func(opts gophercloud.EndpointOpts) (string, error) {
```

Comment:
Same question here, why do we need to change the client here if it’s never used again?